apps/web/src/middlewares/ab-test-middleware.ts

```diff
@@ -0,0 +1,46 @@
+import {
+  EXPERIMENTAL_FEATURE_CONFIGS,
+  FeatureRollOutConfig,
+  ctxKey,
+  ExperimentalFeatureConfigs,
+  EXPERIMENTAL_FEATURES,
+} from 'config/experminetalFeatures'
+import CryptoJS from 'crypto-js'
+import { NextFetchEvent, NextMiddleware, NextResponse } from 'next/server'
+import { MiddlewareFactory, ExtendedNextReq } from './types'
+
+// this function generates a deterministic result for a user for a given feature
+// it hashes a concatination of the users ip together with the features identifier and
+// probanility value. this allows us to ensure that a users probability result is different
+// for each feature gauranteeing a better distribution
+export const getExperimentalFeatureAccessList = (
+  userIdentifier: string,
+  abTestingfeatureFlagInfo: ExperimentalFeatureConfigs,
+): Array<{ feature: EXPERIMENTAL_FEATURES; hasAccess: boolean }> => {
+  const userWhitelistResults = abTestingfeatureFlagInfo.map(
+    (flag: FeatureRollOutConfig): { feature: EXPERIMENTAL_FEATURES; hasAccess: boolean } => {
+      if (flag.whitelist.includes(userIdentifier)) return { feature: flag.feature, hasAccess: true }
+
+      const hash = CryptoJS.SHA256(`${userIdentifier}-${flag.feature}-${flag.percentage}`)
```

Comment:
why're we adding percentage to hash?

Reply:
let's remove it. If percentage increased from 0.05 -> 0.1, the user covered by 0.05 should still be covered under 0.1